src/vm-memory/src/mmap.rs

```diff
@@ -74,31 +109,39 @@ impl Bytes<MemoryRegionAddress> for GuestRegionMmap {
     type E = guest_memory::Error;
 
     fn write(&self, buf: &[u8], addr: MemoryRegionAddress) -> guest_memory::Result<usize> {
-        // TODO: here be write!
         let maddr = addr.raw_value() as usize;
-        self.local_volatile_slice()
+        let bytes = self
+            .local_volatile_slice()
             .write(buf, maddr)
-            .map_err(Into::into)
+            .map_err(Into::<guest_memory::Error>::into)?;
+        self.mark_dirty_pages(maddr, bytes);
+        Ok(bytes)
     }
 
     fn read(&self, buf: &mut [u8], addr: MemoryRegionAddress) -> guest_memory::Result<usize> {
-        // TODO: here be read!
         let maddr = addr.raw_value() as usize;
         self.local_volatile_slice()
             .read(buf, maddr)
             .map_err(Into::into)
     }
 
     fn write_slice(&self, buf: &[u8], addr: MemoryRegionAddress) -> guest_memory::Result<()> {
-        // TODO: here be write!
         let maddr = addr.raw_value() as usize;
-        self.local_volatile_slice()
-            .write_slice(buf, maddr)
-            .map_err(Into::into)
+        match self.local_volatile_slice().write_slice(buf, maddr) {
+            Ok(()) => {
+                self.mark_dirty_pages(maddr, buf.len());
+                Ok(())
+            }
+            Err(e) => {
+                if let VolatileMemoryError::PartialBuffer { completed, .. } = e {
+                    self.mark_dirty_pages(maddr, completed);
+                }
+                Err(e.into())
```

Comment:
We should also do this for `write`, `read_from`, `read_exact_from`.

Reply:
I think `write_slice` is the only one of them  that can return a `PartialBuffer` error.